snapcraft/plugins/rust.py

```diff
@@ -194,21 +195,78 @@ def _get_target(self) -> str:
             )
         return rust_target.format("unknown-linux", "gnu")
 
+    def _project_uses_workspace(self) -> bool:
+        cargo_toml_path = Path(self.builddir, "Cargo.toml")
+
+        try:
+            config = open(cargo_toml_path).read()
+        except FileNotFoundError:
+            raise errors.SnapcraftPluginAssertionError(
+                name=self.name, reason="missing required Cargo.toml in source directory"
+            )
+
+        return "workspace" in toml.loads(config)
+
+    def _install_workspace_artifacts(self) -> None:
+        """Install workspace artifacts."""
+        # Find artifacts in release directory.
+        release_dir = Path(self.builddir, "target", "release")
+
+        # Install binaries to bin/.
+        bins_dir = Path(self.installdir, "bin")
+        bins_dir.mkdir(parents=True, exist_ok=True)
+
+        # Install shared objects to usr/lib/<arch-triplet>.
+        # TODO: Dynamic library support needs to be properly added.
+        # Although we install libraries if we find them, they are most
+        # likely going to be missing dependencies, e.g.:
+        # /home/ubuntu/.cargo/toolchains/stable-x86_64-unknown-linux-gnu/lib/libstd-fae576517123aa4e.so
+        libs_dir = Path(self.installdir, "usr", "lib", self.project.arch_triplet)
+        libs_dir.mkdir(parents=True, exist_ok=True)
+
+        # Cargo build marks binaries and shared objects executable...
+        # Search target directory to get these and install them to the
+        # correct location.
+        for path in release_dir.iterdir():
+            if not os.path.isfile(path):
+                continue
+            if not os.access(path, os.X_OK):
+                continue
+
+            # File is executable, now to determine if bin or lib...
+            if path.name.endswith(".so"):
+                file_utils.link_or_copy(path.as_posix(), libs_dir.as_posix())
+            else:
+                file_utils.link_or_copy(path.as_posix(), bins_dir.as_posix())
+
     def build(self):
         super().build()
 
-        # Write a minimal config.
-        self._write_cargo_config()
+        uses_workspaces = self._project_uses_workspace()
+
+        if uses_workspaces:
```

Comment:
does _write_cargo_config not work with workspaces?

Reply:
ah, i did miss that.  I guess that might cause trouble if trying to cross compile huh...

Reply:
but, the more direct use case is the parallel build count

Reply:
Fixed!